### PR TITLE
Implement offline connection secret sharing

### PR DESCRIPTION
**Pull request by majecty**

### CODE REVIEW

File: network/src/control.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use super::addr::SocketAddr;
```
Comment (sgkim126) [inline network/src/control.rs]:
```rust
+use ctypes::H256;

+use super::addr::SocketAddr;
```

Comment (sgkim126) [reply network/src/control.rs]: https://github.com/CodeChain-io/codechain/wiki/Coding-Style#imports-order

File: rpc/src/v1/traits/net.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use ctypes::H256;
+use std;
```
Comment (sgkim126) [inline rpc/src/v1/traits/net.rs]: Personally, I don't recommend `use std`

File: rpc/src/v1/traits/net.rs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+use jsonrpc_core::Result;
+
+build_rpc_trait! {
+    pub trait Net {
+        # [rpc(name = "net_shareSecret")]
+        fn share_secret(&self, H256, std::net::IpAddr, u16) -> Result<()>;
```
Comment (sgkim126) [inline rpc/src/v1/traits/net.rs]: `::std::net::IpAddr`

File: rpc/src/v1/impls/net.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        }
+    }
+}
+
+impl<NC> Net for NetClient<NC>
+where
+    NC: 'static + NetworkControl + Send + Sync,
+{
+    fn share_secret(&self, secret: H256, address: std::net::IpAddr, port: u16) -> Result<()> {
+        let network_control = self.network_control.as_ref().ok_or(errors::network_disabled())?;
```
Comment (sgkim126) [inline rpc/src/v1/impls/net.rs]:
```rust
let network_control = self.network_control.as_ref().ok_or_else(errors::network_disabled)?;
```

File: rpc/src/v1/impls/net.rs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        network_control.register_secret(secret, SocketAddr::new(address, port));
+        Ok(())
+    }
+
+    fn connect(&self, address: std::net::IpAddr, port: u16) -> Result<()> {
+        let network_control = self.network_control.as_ref().ok_or(errors::network_disabled())?;
```
Comment (sgkim126) [inline rpc/src/v1/impls/net.rs]:
```rust
let network_control = self.network_control.as_ref().ok_or_else(errors::network_disabled)?;
```

File: codechain/main.rs [hunk 6]
```diff
@@ -288,6 +269,30 @@ fn run_node(matches: ArgMatches) -> Result<(), String> {
         }
     };
 
+    let rpc_apis_deps = Arc::new(rpc_apis::ApiDependencies {
+        client: client.client(),
+        miner: miner.clone(),
```
Comment (sgkim126) [inline codechain/main.rs]: `Arc::clone(&miner)`

File: codechain/main.rs [hunk 7]
```diff
@@ -288,6 +269,30 @@ fn run_node(matches: ArgMatches) -> Result<(), String> {
...
+        network_control: network_service.map(|network_service| Arc::new(network_service)),
```
Comment (sgkim126) [inline codechain/main.rs]: network_control: network_service.map(Arc::new),

File: codechain/main.rs [hunk 8]
```diff
@@ -288,6 +269,30 @@ fn run_node(matches: ArgMatches) -> Result<(), String> {
...
+    });
+
+    let _rpc_server = {
+        if !config.rpc.disable {
+            let rpc_config = (&config.rpc).into();
+            Some(rpc_start(rpc_config, rpc_apis_deps.clone())?)
```
Comment (sgkim126) [inline codechain/main.rs]: `Some(rpc_start(rpc_config, Arc::clone(&rpc_apis_deps))?)`

File: codechain/main.rs [hunk 9]
```diff
@@ -288,6 +269,30 @@ fn run_node(matches: ArgMatches) -> Result<(), String> {
...
+        } else {
+            None
+        }
+    };
+
+    let _ipc_server = {
+        if !config.ipc.disable {
+            let ipc_config = (&config.ipc).into();
+            Some(rpc_ipc_start(ipc_config, rpc_apis_deps.clone())?)
```
Comment (sgkim126) [inline codechain/main.rs]: `Some(rpc_ipc_start(ipc_config, Arc::clone(&rpc_apis_deps))?)`

File: codechain/main.rs (hunk 6 above)
Comment (sgkim126) [reply codechain/main.rs]: https://doc.rust-lang.org/std/sync/struct.Arc.html#cloning-references

File: network/src/routing_table.rs [hunk 10]
```diff
@@ -172,9 +253,25 @@ impl RoutingTable {
...
             let entry = entry.lock();
             let old_state = entry.replace(State::Intermediate);
-            if let State::SecretShared(shared_secret) = &old_state {
+            let shared_secret = match old_state {
+                State::SecretShared(shared_secret) => Some(shared_secret),
+                State::SecretPreimported(secret) => Some(secret),
+                _ => None,
+            };
+
+            let secret_origin = match old_state.clone() {
```
Comment (sgkim126) [inline network/src/routing_table.rs]: `rust
let secret_origin = match &old_state {
`

File: network/src/routing_table.rs [hunk 11]
```diff
@@ -172,9 +253,25 @@ impl RoutingTable {
...
+                State::SecretShared(_secret) => Some(SecretOrigin::Shared),
```
Comment (sgkim126) [inline network/src/routing_table.rs]: `rust
State::SecretShared(_)
`

File: network/src/routing_table.rs [hunk 12]
```diff
@@ -172,9 +253,25 @@ impl RoutingTable {
...
+                State::SecretPreimported(_secret) => Some(SecretOrigin::Preimported),
```
Comment (sgkim126) [inline network/src/routing_table.rs]:
```rust
State::SecretPreimported(_) => Some(SecretOrigin::Preimported),
```

**Comment by sgkim126** [comment]
LGTM about the last commit. Let's wait for @joojis review RPC document.

File: rpc/src/v1/impls/net.rs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use super::super::errors;
+use super::super::traits::Net;
+use cnetwork::{NetworkControl, SocketAddr};
+use ctypes::H256;
+use jsonrpc_core::Result;
+use std;
+use std::sync::Arc;
```
Comment (joojis) [inline rpc/src/v1/impls/net.rs]: Please fix the order of imports.

https://github.com/CodeChain-io/codechain/wiki/Coding-Style#imports-order

**Comment by sgkim126** [comment]
Fixes #147

**Pull request merged**